### ISSUE TITLE
Binary NRRD import/export of default data

### ISSUE THREAD

**Issue by sebastianuber** [e0404/matRad#282]
Hello,

I played around with the binary import and export in NRRD format of the default datasets and noticed the following.

The binary export of a plan with a default dataset (I used prostate) produced the following error;

Error using  +

Matrix dimensions must agree.

Error in matRad_writeCube (line 75)

    metadata.imageCenter = metadata.imageOrigin + (imageExtent/2);

Error in matRad_exportGUI>btn_export_Callback (line 413)

    matRad_writeCube(fullfile(exportDir,['CT_ED' extension]),ct.cube{1},'double',metadata);

Error in gui_mainfcn (line 95)

        feval(varargin{:});

Error in matRad_exportGUI (line 55)

    gui_mainfcn(gui_State, varargin{:});

Error in

matlab.graphics.internal.figfile.FigFile/read>@(hObject,eventdata)matRad_exportGUI('btn_export_Callback',hObject,eventdata,guidata(hObject))

Error while evaluating UIControl Callback

The export worked after I added a transpose at line 73 in writeCube;

if ~isfield(metadata,'imageCenter')

    % edit by Sebastian; fixed by adding transpose?

    metadata.imageCenter = metadata.imageOrigin.' + (imageExtent/2);

    % metadata.imageCenter = metadata.imageOrigin + (imageExtent/2);

end

 

I am not sure if this is the right fix and if it messes up anything else.

Then there is an error with the import of binary data;

Error Dialog: Binary Patient Import: Could not import data

Reconversion of HU values could not be done because HLUT is not bijective.

-1024.0 0.00324

200.0   1.20000

449.0   1.20000 (change to 1.20001?)

2000.0  2.49066    

2048    2.53060

3071    2.53060 (change to 2.53061?)

 

Excerpt from the default HLUTs. The identical values seem to be the cause of the error. Should the default HLUTs maybe be changed so that this error doesn’t occur naturally?

Kind regards,
Sebastian Uber
MSc Student Oncoray Dresden

**Comment by wahln** [comment]
Hi Sebastian, 
thanks for the detailed bug report. Are you using the import/export interface from the GUI or are you calling the routine through a script? 
Cheers, Niklas

**Comment by sebastianuber** [comment]
Hello Niklas,

you're welcome. I was using the interface from the GUI.

Kind regards,
Sebastian

**Comment by wahln** [comment]
Three things:

1. I can't reproduce it with the current dev and master branch and default data since with the default data the fields you mentioned should be obtained from dicomInfo stored in the ct structure field.
2. it's a bug nevertheless, since as soon the dicomInfo fields are not populated these values are not computed correctly.  I think your fix should be fine for now and not break some other things. 
3. The default HLUT should now be changed (on the dev branch) to be bijective.

So in summary: I will prepare a bugfix which handles everything correctly, but for now your fixes should work without any problems. And you could check if you are using the current version of the default data.

Cheers, Niklas

**Comment by sebastianuber** [comment]
That is very good to hear. I do have the current default data.

Cheers

**Comment by wahln** [comment]
I prepared a pull request that should hopefully fix this issue / get up to date with the current cube and HLUT handling, see #288 

In particular 

- imageOrigin and imageExtent should now always be row vectors.
- when importing a ct the HLUT can be selected (or no conversion can be performed)
